dashboard/app/models/contact_rollups_pardot_memory.rb

```diff
@@ -31,50 +31,89 @@ def self.add_and_update_pardot_ids(last_id = nil)
+    pardot_writer = PardotV2.new
+    ActiveRecord::Base.connection.exec_query(query_updated_contacts).each do |record|
+      # If pardot_id has changed since the last data sync, we should assume that
+      # Pardot prospect data is currently empty and re-sync all contact data.
+      old_prospect_data =
+        record['pardot_id_changed'] ?
+          {} :
+          JSON.parse(record['data_synced'] || '{}').deep_symbolize_keys
+      new_contact_data = JSON.parse(record['data']).deep_symbolize_keys
+
+      submissions, errors = pardot_writer.batch_update_prospects(
+        record['email'],
+        record['pardot_id'],
+        old_prospect_data,
+        new_contact_data
+      )
+      save_sync_results(submissions, errors, Time.now.utc) if submissions.present?
+    end
+
+    submissions, errors = pardot_writer.batch_update_remaining_prospects
+    save_sync_results(submissions, errors, Time.now.utc) if submissions.present?
+  end
+
+  def self.query_new_contacts
     # Finds contacts in contact_rollups_processed but are not in contact_rollups_pardot_memory.
     # Contacts must not be previously rejected by Pardot as invalid emails.
-    new_contacts_query = <<-SQL.squish
+    <<-SQL.squish
       SELECT processed.email, processed.data
       FROM contact_rollups_processed AS processed
       LEFT OUTER JOIN contact_rollups_pardot_memory AS pardot
         ON processed.email = pardot.email
       WHERE pardot.pardot_id IS NULL
         AND NOT (pardot.data_rejected_reason <=> '#{PardotHelpers::ERROR_INVALID_EMAIL}')
     SQL
+  end
 
-    # Adds contacts to a batch and then sends batch requests to create new Pardot prospects.
-    # Requests may not be sent immediately until batch size is big enough.
-    pardot_writer = PardotV2.new
-
-    ActiveRecord::Base.connection.exec_query(new_contacts_query).each do |record|
-      data_col = JSON.parse(record['data']).deep_symbolize_keys
-      contact = {email: record['email']}.merge(data_col)
-
-      submissions, errors = pardot_writer.batch_create_prospects contact
-      save_sync_results(submissions, errors, Time.now) if submissions.present?
-    end
-
-    # There could be prospects left in the batch because batch size is not yet big enough
-    # to trigger a Pardot request. Sends the remaining of the batch to Pardot now.
-    submissions, errors = pardot_writer.batch_create_remaining_prospects
-    save_sync_results(submissions, errors, Time.now) if submissions.present?
+  def self.query_updated_contacts
+    <<-SQL.squish
+      SELECT
+        processed.email, processed.data,
+        pardot.pardot_id, pardot.data_synced,
+        (pardot.pardot_id_updated_at > pardot.data_synced_at) AS pardot_id_changed
+      FROM contact_rollups_processed AS processed
+      INNER JOIN contact_rollups_pardot_memory AS pardot
+        ON processed.email = pardot.email
+      WHERE pardot.pardot_id IS NOT NULL
+        AND (
+          (pardot.data_synced_at IS NULL)
```

Comment:
Maybe  note here describing what each of these three cases represents? To see if I understand correctly, I'll try and write them -- they are:
1. ...I'm actually not sure on this one, I thought this would be covered by the new contact sync step?
2. A row containing data for that contact has been updated in our database since that contact was last synced to Pardot
3. The contact's Pardot ID has been updated since last time contact rollups process ran.

Reply:
(1) means this contact has never been synced to Pardot. Another way to see it is that (1) is a special case of (2). (2) doesn't work if `pardot.data_synced_at` is null. I will add comments to explain these conditions.